pkg/fileutil/lock_windows.go

```diff
@@ -33,39 +52,131 @@ type Lock interface {
 	Destroy() error
 }
 
-type lock struct {
-	fd   int
-	file *os.File
+// fileMutex is similar to sync.RWMutex, but also synchronizes across processes.
+// This implementation is based on flock syscall.
+// https://github.com/golang/build/blob/master/cmd/builder/filemutex_windows.go
+type fileMutex struct {
+	locked   int32 // prevent deadlock in user memory-level
+	mu       sync.RWMutex
+	fd       syscall.Handle
+	filename string
 }
 
-func (l *lock) Name() string {
-	return l.file.Name()
+func lockFileEx(h syscall.Handle, flags, reserved, locklow, lockhigh uint32, ol *syscall.Overlapped) (err error) {
+	r1, _, e1 := syscall.Syscall6(procLockFileEx.Addr(), 6, uintptr(h), uintptr(flags), uintptr(reserved), uintptr(locklow), uintptr(lockhigh), uintptr(unsafe.Pointer(ol)))
+	if r1 == 0 {
+		if e1 != 0 {
+			err = error(e1)
+		} else {
+			err = syscall.EINVAL
+		}
+	}
+	return
 }
 
-// TryLock acquires exclusivity on the lock without blocking
-func (l *lock) TryLock() error {
-	return nil
+func unlockFileEx(h syscall.Handle, locklow, lockhigh uint32, ol *syscall.Overlapped) (err error) {
+	var reserved uint32 = 0
+	r1, _, e1 := syscall.Syscall6(procUnlockFileEx.Addr(), 5, uintptr(h), uintptr(reserved), uintptr(locklow), uintptr(lockhigh), uintptr(unsafe.Pointer(ol)), 0)
+	if r1 == 0 {
+		if e1 != 0 {
+			err = error(e1)
+		} else {
+			err = syscall.EINVAL
+		}
+	}
+	return
 }
 
-// Lock acquires exclusivity on the lock without blocking
-func (l *lock) Lock() error {
+// NewLock creates a NewLock.
+func NewLock(file string) (Lock, error) {
+	if file == "" {
+		return &fileMutex{fd: syscall.InvalidHandle, filename: file}, fmt.Errorf("cannot open empty filename")
+	}
+	fd, err := syscall.CreateFile(&(syscall.StringToUTF16(file)[0]),
+		syscall.GENERIC_READ|syscall.GENERIC_WRITE,
+		syscall.FILE_SHARE_READ|syscall.FILE_SHARE_WRITE|syscall.FILE_SHARE_DELETE,
+		nil,
+		syscall.OPEN_ALWAYS,
+		syscall.FILE_ATTRIBUTE_NORMAL,
+		0)
+	if err != nil {
+		return nil, err
+	}
+	return &fileMutex{fd: fd, filename: file}, nil
+}
+
+// TryLock acquires exclusivity on the lock without blocking.
+func (fm *fileMutex) TryLock() error {
+	return fm.lockFile(LOCKFILE_FAIL_IMMEDIATELY)
+}
+
+// Lock acquires exclusivity on the lock.
+// It blocks until the lock is acquired.
+// https://msdn.microsoft.com/en-us/library/windows/desktop/aa365202(v=vs.85).aspx
+func (fm *fileMutex) Lock() error {
+	return fm.lockFile(0)
+}
+
+// Unlock unlocks the lock.
+func (fm *fileMutex) Unlock() error {
+	if fm.fd != syscall.InvalidHandle {
+		if err := unlockFileEx(fm.fd, 1, 0, &syscall.Overlapped{}); err != nil {
+			return err
+		}
+	}
+	fm.mu.Unlock()
+	atomic.StoreInt32(&fm.locked, 0)
 	return nil
 }
 
-// Unlock unlocks the lock
-func (l *lock) Unlock() error {
+func (fm *fileMutex) lockFile(flags uint32) error {
+	if atomic.LoadInt32(&fm.locked) == 1 {
+		return ErrLocked
+	}
+	atomic.StoreInt32(&fm.locked, 1)
+	fm.mu.Lock()
```

Comment:
if `lockFileEx` gives `err != nil`, `lockFile` will exit with an error still holding the mutex; if something else tries to `Lock`, there'll be a deadlock